backends/oauth2.go

```diff
@@ -0,0 +1,335 @@
+package backends
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"github.com/iegomez/mosquitto-go-auth/backends/topics"
+	"github.com/pkg/errors"
+	log "github.com/sirupsen/logrus"
+	go_oauth2 "golang.org/x/oauth2"
+	go_clientcredentials "golang.org/x/oauth2/clientcredentials"
+	"net/http"
+	"strconv"
+	"strings"
+	"time"
+)
+
+type userState struct {
+	username          string
+	superuser         bool
+	readTopics        []string
+	writeTopics       []string
+	lastUserInfoUpate time.Time
+	createdAt         time.Time
+	updatedAt         time.Time
+	client            *http.Client
+	token             *go_oauth2.Token
+}
+
+type UserInfo struct {
+	sub  string `json:"sub"`
+	MQTT struct {
+		Topics struct {
+			Read  []string `json:"read"`
+			Write []string `json:"write"`
+		} `json:"topics"`
+		Superuser bool `json:"superuser"`
+	} `json:"mqtt"`
+}
+
+type Oauth2 struct {
+	oauth2Config            go_oauth2.Config
+	clientcredentialsConfig go_clientcredentials.Config
+	tokenUrl                string
+	userInfoURL             string
+	userCache               map[string]userState
```

Comment:
Actually I was not yet aware of the general cache.

If I'd remove the specialized cache the plugin could make more requests to the oauth2-server than necessary, as far as I can tell. This happens if one client wants to read or write from multiple topics because the plugin would send a request to the server for every topic individually.

In contrast the specialized cache makes a single request and stores access permissions for all topics. If the same clients requests access for another topic that cache will be used.

Is there a way to reproduce this behavior using the general cache?

Reply:
Ah, indeed, you're caching the whole user information here, not just some read/written/subscribed topic.
I'll give it some thought and get back to you.

Reply:
Hey @iegomez, did you reach a decision?

Reply:
Sorry, I've been quite busy lately and haven't checked the plugin in a while. I'll try to go over the PR again this week.